src/services/HashGeneratorService.php

```diff
@@ -64,17 +66,20 @@ protected static function buildDataString(InvoiceRecord $record)
                 'FechaHoraHusoGenRegistro=' . trim($fields['recordTimestamp']),
             ];
             return implode('&', $parts);
-
-        } elseif ($record instanceof \eseperio\verifactu\models\InvoiceCancellation) {
+        }
+        // Detect type: submission or cancellation
+        if ($record instanceof InvoiceCancellation) {
+            // Get invoice ID and chaining using getter methods
+            $invoiceId = $record->getInvoiceId();
+            $chaining = $record->getChaining();
             // Fields order for "RegistroAnulacion" (cancellation)
             $fields = [
-                'issuerNif'         => $record->invoiceId->issuerNif,
-                'seriesNumber'      => $record->invoiceId->seriesNumber,
-                'issueDate'         => $record->invoiceId->issueDate,
-                'hash'              => isset($record->chaining['previousHash']) ? $record->chaining['previousHash'] : '',
+                'issuerNif'         => $invoiceId->issuerNif,
+                'seriesNumber'      => $invoiceId->seriesNumber,
+                'issueDate'         => $invoiceId->issueDate,
+                'hash'              => $chaining && $chaining->getPreviousInvoice() ? $chaining->getPreviousInvoice()->hash : '',
```

Comment:
This could be null safe operator and null coallesce `$chaining?->getPreviousInvoice() ?? ''`